### PR TITLE
[OTA-2551] Move to {Device,Ecu}UpdateAssignment{Repository}

### PR DESCRIPTION
**Pull request by jerrytrieu**
Signed-off-by: Jerry Trieu <jerry.trieu@here.com>

### CODE REVIEW

File: src/main/scala/com/advancedtelematic/director/data/DataType.scala [hunk 1]
```diff
@@ -59,6 +59,20 @@ object DataType {
...
+                                   version: Int,
+                                   customImage: CustomImage)
+
+  final case class DeviceUpdateAssignment(
+    namespace: Namespace,
+    deviceId: DeviceId,
+    correlationId: Option[CorrelationId],
+    updateId: Option[UpdateId],
+    version: Int,
+    inFlight: Boolean)
```
Comment (houcros) [inline src/main/scala/com/advancedtelematic/director/data/DataType.scala]: Why not use the same field name than in the table here? IMO `served` sounds more understandable.

File: src/main/scala/com/advancedtelematic/director/db/DeviceUpdateAssignmentRepository.scala [hunk 2]
```diff
@@ -0,0 +1,113 @@
...
+
+protected class DeviceUpdateAssignmentRepository()(implicit db: Database, ec: ExecutionContext) {
+  import Schema.deviceUpdateAssignments
+
+  def exists(namespace: Namespace, deviceId: DeviceId, version: Int): Future[Boolean] =
+    db.run(existsAction(namespace, deviceId, version))
+
+  protected [db] def existsAction(namespace: Namespace, deviceId: DeviceId, version: Int)
+  : DBIO[Boolean] =
+    deviceUpdateAssignments
```
Comment (houcros) [inline src/main/scala/com/advancedtelematic/director/db/DeviceUpdateAssignmentRepository.scala]: Not using the `namespace` to filter?

File: src/main/scala/com/advancedtelematic/director/db/DeviceUpdate.scala [hunk 3]
```diff
@@ -13,16 +13,18 @@ object DeviceUpdateResult {
   sealed abstract class DeviceUpdateResult
 
   final case object NoUpdate extends DeviceUpdateResult
-  final case class UpdateNotCompleted(updateTarget: DeviceUpdateTarget) extends DeviceUpdateResult
-  final case class UpdateSuccessful(updateTarget: DeviceUpdateTarget) extends DeviceUpdateResult
+  final case class UpdateNotCompleted(updateTarget: DeviceUpdateAssignment) extends DeviceUpdateResult
```
Comment (houcros) [inline src/main/scala/com/advancedtelematic/director/db/DeviceUpdate.scala]: Maybe we could also change the field name to `updateAssignment`? Or will that propagate to other services and get messy?

File: src/main/scala/com/advancedtelematic/director/db/DeviceUpdateAssignmentRepository.scala [hunk 4]
```diff
@@ -0,0 +1,113 @@
...
+    deviceId: DeviceId,
+    correlationId: Option[CorrelationId],
+    updateId: Option[UpdateId])
+  : DBIO[Int] = for {
+    version <- fetchLatestAction(namespace, deviceId).asTry.flatMap {
+      case Success(x) => DBIO.successful(x)
+      case Failure(NoTargetsScheduled) => DBIO.successful(0)
+      case Failure(ex) => DBIO.failed(ex)
+    }
+    new_version = version + 1
```
Comment (houcros) [inline src/main/scala/com/advancedtelematic/director/db/DeviceUpdateAssignmentRepository.scala]: Camel case?

File: src/main/scala/com/advancedtelematic/director/db/DeviceUpdateAssignmentRepository.scala [hunk 5]
```diff
@@ -0,0 +1,113 @@
...
+      .filter(_.version > fromVersion)
+      .sortBy(_.version)
+      .result
+
+  def getUpdatesFromTo(namespace: Namespace, deviceId: DeviceId, fromVersion: Int, toVersion: Int)
+  : Future[Seq[DeviceUpdateAssignment]] = db.run {
+    Schema.deviceUpdateAssignments
+      .filter(_.namespace === namespace)
+      .filter(_.deviceId === deviceId)
+      .filter(_.version > fromVersion)
```
Comment (houcros) [inline src/main/scala/com/advancedtelematic/director/db/DeviceUpdateAssignmentRepository.scala]: Would it make sense to have the inclusion in the lower bound instead? Or is this actually the use we need for this method? Slicing like this is usually the interval `[from, to)`, e. g. Java `String#substring`, so doing it the other way around might be unexpected to the user of this method.

File: src/main/scala/com/advancedtelematic/director/db/EcuUpdateAssignmentMigration.scala [hunk 6]
```diff
@@ -0,0 +1,60 @@
...
+
+  private val _log = LoggerFactory.getLogger(this.getClass)
+
+  implicit val getResult: GetResult[Namespace] = pr => {
+    implicitly[ColumnType[Namespace]].getValue(pr.rs, 1)
+  }
+
+  def run: Future[Done] = {
+    val source = db.stream(Schema.deviceTargets.result)
+    Source.fromPublisher(source).mapAsyncUnordered(1) { case oldAssignment =>
```
Comment (houcros) [inline src/main/scala/com/advancedtelematic/director/db/EcuUpdateAssignmentMigration.scala]: `case` not needed here.

File: src/main/scala/com/advancedtelematic/director/db/EcuUpdateAssignmentMigration.scala [hunk 7]
```diff
@@ -0,0 +1,60 @@
...
+      val queryNamespace = sql"""SELECT DISTINCT(namespace) from `ecus` WHERE device = '#${oldAssignment.device}'""".as[Namespace]
+      db.run(queryNamespace).flatMap { namespace =>
```
Comment (houcros) [inline src/main/scala/com/advancedtelematic/director/db/EcuUpdateAssignmentMigration.scala]: Isn't there a way to do this with a single `db.run`? Maybe something similar to this (I haven't tested it)?:
```
val source = db.stream(Schema.deviceTargets.result)
    Source.fromPublisher(source).mapAsyncUnordered(1) { oldAssignment =>
      val queryNamespaces = sql"""SELECT DISTINCT(namespace) from `ecus` WHERE device = '#${oldAssignment.device}'""".as[Namespace]
      db.run {
        for {
          namespaces <- queryNamespaces
          assignment <- deviceUpdateAssignmentRepository.persistAction(
            namespaces.head,
            oldAssignment.device,
            oldAssignment.correlationId,
            oldAssignment.updateId,
            oldAssignment.targetVersion)
          targets <- ecuUpdateAssignmentRepository.legacyFetchAction(
            assignment.namespace,
            assignment.deviceId,
            assignment.version)
          _ <- ecuUpdateAssignmentRepository.persistAction(
            assignment.namespace,
            assignment.deviceId,
            assignment.version,
            targets)
        } yield ()
      }
    }.runWith(Sink.foreach { _ => () })
```

File: src/main/resources/db/migration/V22__device_update_assignments.sql [hunk 8]
```diff
@@ -0,0 +1,28 @@
+CREATE TABLE `device_update_assignments` (
```
Comment (simao) [inline src/main/resources/db/migration/V22__device_update_assignments.sql]: Why do we need both device update assignments and ecu update assignments? device assignments could be inferred from ecu assignments?

File: src/main/scala/com/advancedtelematic/director/db/EcuUpdateAssignmentMigration.scala [hunk 9]
```diff
@@ -0,0 +1,61 @@
...
+  with EcuUpdateAssignmentRepositorySupport {
+
+  private val _log = LoggerFactory.getLogger(this.getClass)
+
+  implicit val getResult: GetResult[Namespace] = pr => {
+    implicitly[ColumnType[Namespace]].getValue(pr.rs, 1)
+  }
+
+  def run: Future[Done] = {
+    val source = db.stream(Schema.deviceTargets.result)
```
Comment (simao) [inline src/main/scala/com/advancedtelematic/director/db/EcuUpdateAssignmentMigration.scala]:
```sql
select dt.*, e.namespace from device_update_targets dt inner join ecus e on e.device = dt.device
```
But anyway not really important

Comment (jerrytrieu) [reply src/main/scala/com/advancedtelematic/director/db/EcuUpdateAssignmentMigration.scala]: How to get a `DeviceUpdateTarget` out of that query?

Comment (simao) [reply src/main/scala/com/advancedtelematic/director/db/EcuUpdateAssignmentMigration.scala]: You'd need to write a `GetResult[(DeviceUpdateTarget, Namespace)]`

File: src/main/resources/db/migration/V22__device_update_assignments.sql (hunk 8 above)
Comment (jerrytrieu) [reply src/main/resources/db/migration/V22__device_update_assignments.sql]: Device update assignment has correlationId per update and multiple Ecu update assignment has target per ecu. Can try simplify later.

**Comment by jerrytrieu** [comment]
Passes e2e on berlin hat.

**Pull request merged**